tools/osbuild-mpp

```diff
@@ -1098,6 +1099,18 @@ class ManifestFile:
         else:
             return fakeroot[0], False
 
+    def _format_dict_node(self, node, stack):
+        if len(stack) > 0:
+            parent_node = stack[-1][0]
+            parent_key = stack[-1][1]
+        else:
+            parent_node = None
+            parent_key = None
+
+        if parent_key == "stages":
+            pipeline_name = self.get_pipeline_name(parent_node)
+            self._process_stage(node, pipeline_name)
+
```

Comment:
So any dict with a key called `"stages"` now gets the stage-preprocessor run? I guess this works because the stage-preprocessor only reacts to mpp keys, and you better know where you place them. But... ummm. At this point I don't understand why to distinguish them at all from the format nodes?

Reply:
> But... ummm. At this point I don't understand why to distinguish them at all from the format nodes?

That is a good question, as I initially wanted to do that. However, there are some ordering constraints and differences that make this complex. For example, take `mpp-embed` used in an stage. It will generate the input source and set a variable with its name, like this:

```
  stages:
  - type: org.osbuild.copy
    inputs:
      inlinefile:
        type: org.osbuild.files
        origin: org.osbuild.source
        mpp-embed:
          id: luks-key
          text: $luks_passphrase
    options:
      paths:
        - from:
            mpp-format-string: input://inlinefile/{embedded['luks-key']}
          to: tree:///luks-key
```

There are two issues with this. First of all we have to be able to use the variable generated by the mpp-embed in the options, so we can't wait until the entire stage is formated. To make this work i had to run the process_stage before formating the stage (which matches what happened before, but seems a bit weird).

Secondly, the format of the "mpp-embed" is different than the formating nodes. It is part of (a key in) an existing input, rather than a completely separate mpp dict that will be replaced.

That doesn't mean it is impossible to handle them more in-line with the other formating keys, but its not quite as natural as you'd think.

Reply:
Right. That explains why we keep them split. Thanks!

But why guard the stage handling by the `key == "stages"` protection? I don't mind it too much, but we could also just call it on *any* entry, right? It is basically just a safety guard to prevent you from doing stupid things too easily?

I don't mind it too much, just trying to understand.

Reply:
Also, I initially wanted to make the "is it a stage" detection a bit more precise. For example, we could (in the V2 version at least) just check if the parent is self.pipelines. But I guess that got dropped at some point. Do you think that would be useful?

Reply:
> Also, I initially wanted to make the "is it a stage" detection a bit more precise. For example, we could (in the V2 version at least) just check if the parent is self.pipelines. But I guess that got dropped at some point. Do you think that would be useful?

But then you need to move it into the derived classes. I would rather get rid of the v1-formatter.

So how about just adding a comment there? Something like "Avoid running the stage formatter on anything that does not look like a stage"?

If you rather extract it and check for "parent == self.pipeline", please do! But I am ok with a comment as well.

Reply:
In general, how important is the V1 generator really? Seems like dropping that at some point would clean up some stuff.

Reply:
Added a comment